### PR TITLE
Add basic Charlieplexing scan

### PR DESCRIPTION
**Pull request by kienvo**
Hey, I just finished some basic initial code for the project.
There are some problems with the interrupt when using xpack's `riscv-none-elf-gcc`:

- The gpio interrupt handler only runs once
- The timer interrupt handler is so slow

Switching back to `riscv-none-embed-gcc` seems working, but the timer interrupt handler takes so much time that the main loop doesn't have enough time slice to execute.

Main changes:

- Generate code from SDK
- Add basic Charlieplexing scan

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -50,6 +50,7 @@ CH5xx_ble_firmware_library/StdPeriphDriver/CH58x_usb2hostBase.c \
 CH5xx_ble_firmware_library/StdPeriphDriver/CH58x_spi1.c \
 CH5xx_ble_firmware_library/RVMSIS/core_riscv.c \
 src/main.c \
+leddrv.c
```
Comment (fcartegnie) [inline Makefile]: path issue

Comment (kienvo) [reply Makefile]: My bad. But somehow the build still works. 

I'll get it fixed right away.

File: src/leddrv.c [hunk 2]
```diff
@@ -3,46 +3,61 @@
 #define LED_DRIVE_STRENTH GPIO_ModeOut_PP_20mA
 #define LED_PINCOUNT (23)
 
+#define PA_BASE 	((uint32_t *)0x400010A0)
```
Comment (fcartegnie) [inline src/leddrv.c]: I don't know if we are going to access non 4 bytes aligned offsets, but in that case, it will deref the wrong address

Comment (kienvo) [reply src/leddrv.c]: There is no access to non 4 bytes aligned offsets. But I've just found pre-defined base address in the StdPeriphDriver; I'll use that instead.

Comment (fcartegnie) [reply src/leddrv.c]: I see you just changed the code, so I guess you understood the issue with gpio_base being already uint32_t aligned

`*(uint32_t *)(gpio_base + GPIO_PD_DRV) = *cfg;`

Comment (kienvo) [reply src/leddrv.c]: I had a workaround for that before the last change:
```
uint32_t b = (uint32_t)gpio_base; // <--here
*(uint32_t *)(b + GPIO_PD_DRV) = *cfg;
```
It seems kind of ugly but I changed to use pre-defined base address in the last change anyway.

**Review by fcartegnie — CHANGES_REQUESTED** [review]
that is not matching coding style and unneeded

**Comment by kienvo** [comment]
> that is not matching coding style and unneeded

Did you mean commit 817425b? It seem nonsense but for quick testing. I'll remove it here and add changing brightness feature in another PR. How do you think?

**Comment by fcartegnie** [comment]
Makefile changes

**Comment by kienvo** [comment]
> Makefile changes

954f786
4cece4e
d812e56
208b854
Which one?

**Comment by fcartegnie** [comment]
> > Makefile changes
> 
> [4cece4e](https://github.com/fossasia/badgemagic-firmware/commit/4cece4e55af7ba2a0568a14c8b917cd48341095c)

**Comment by kienvo** [comment]
What coding style do you suggest?

**Comment by kienvo** [comment]
I see many projects use the old style, but isn't it convenient to have this appending style? It could be 
- turned on/off a flag by commenting out a line 
- or add an end-line comment without affecting other flags.

**Comment by kienvo** [comment]
@fcartegnie if there is no other change I think it is time to get this PR merge.

**Comment by fcartegnie** [comment]
> @fcartegnie if there is no other change I think it is time to get this PR merge.

You can't merge do & undos or fixs to those commits.

Please rebase -i and discard reverted and reverter code, and merge intra fix changes

**Comment by kienvo** [comment]
> You can't merge do & undos or fixs to those commits.
> 
> Please rebase -i and discard reverted and reverter code, and merge intra fix changes

Then force push or should I make a new PR?

**Comment by fcartegnie** [comment]
> > You can't merge do & undos or fixs to those commits.
> > Please rebase -i and discard reverted and reverter code, and merge intra fix changes
> 
> Then force push or should I make a new PR?

use the force, luke

**Review by fcartegnie — APPROVED** [review]
unsure why we need to commit so much code / project template

**Comment by kienvo** [comment]
I could write a script to automatically download and generate `StdPeriphDriver` if this is considered as the optimal way.

**Comment by simon-budig** [comment]
Please keep StdPeriphDriver in the repository. I believe it would be a bad idea to have our repository depend on Code pulled in at build-time from an external source. That creates a dependency we don't want.

(admittedly I don't like the StdPeriphDriver code. I believe at some point we might want to fix/modify it. It is inconsistent within itself and has some stupid design descisions. Being able to fix it is an important reason to keep the code in our repository.)

**Pull request merged**